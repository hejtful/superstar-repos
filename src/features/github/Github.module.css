.main {
  height: 100%;
}

.banner {
  background-color: rgb(22, 27, 34);
  border-bottom: 1px solid rgb(48, 54, 61);
  padding: 40px;
  text-align: center;
}

.bannerTitle {
  color: rgb(201, 209, 217);
  font-size: 32px;
  font-weight: bold;
  line-height: 1.5;
  margin: 0;
}

.bannerDescription {
  color: rgb(139, 148, 158);
  font-size: 16px;
  line-height: 1.5;
  margin: 0 0 10px;
}

.container {
  background-color: rgb(13, 17, 23);
  padding: 20px;
}

.wrapper {
  background-color: rgb(13, 17, 23);
  border: 1px solid rgb(48, 54, 61);
  border-radius: 6px;
  max-width: 800px;
  margin: 0 auto;
}

.actions {
  background-color: rgb(22, 27, 34);
  border-bottom: 1px solid rgb(48, 54, 61);
  display: flex;
  justify-content: center;
  padding: 16px;
}

.statusMessage {
  padding: 16px;
  color: rgb(139, 148, 158);
}
