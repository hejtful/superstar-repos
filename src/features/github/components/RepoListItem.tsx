import { MouseEvent, EventHandler } from 'react';
import { Repository } from '../githubTypes';
import styles from './RepoListItem.module.css';

interface Props {
  repo: Repository;
  isStarred: boolean;
  onStarButtonClick: EventHandler<MouseEvent>;
  onUnStarButtonClick: EventHandler<MouseEvent>;
}

export function RepoListItem({
  repo,
  isStarred,
  onStarButtonClick,
  onUnStarButtonClick,
}: Props) {
  return (
    <article className={styles.wrapper} data-testid="repo-list-item">
      <div className={styles.titleWrapper}>
        <h1 className={styles.title}>
          <a href={repo.html_url} className={styles.link}>
            {repo.owner.login}
            <span className={styles.repoName}> / {repo.name}</span>
          </a>
        </h1>

        {isStarred ? (
          <button
            className={styles.button}
            type="button"
            onClick={onUnStarButtonClick}
          >
            Un-star
          </button>
        ) : (
          <button
            className={styles.button}
            type="button"
            onClick={onStarButtonClick}
          >
            Star
          </button>
        )}
      </div>

      <p className={styles.description}>{repo.description}</p>

      <div className={styles.details}>
        <span>Language: {repo.language || 'Unknown'}</span>
        <span>Stars: {repo.stargazers_count}</span>
      </div>
    </article>
  );
}
