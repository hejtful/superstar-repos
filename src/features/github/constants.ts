export const languagesPerPage = 25;

export const languages = [
  'JavaScript',
  'Python',
  'Java',
  'PHP',
  'C++',
  'C',
  'C#',
  'Ruby',
  'Scala',
  'Haskell',
  'Shell',
  'Elm',
  'Rust',
  'Kotlin',
];
