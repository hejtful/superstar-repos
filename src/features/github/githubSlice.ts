import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../app/store';

export interface GithubState {
  starred: number[];
}

const initialState: GithubState = {
  starred: [],
};

export const githubSlice = createSlice({
  name: 'github',
  initialState,
  reducers: {
    starRepo: (state, action: PayloadAction<number>) => {
      state.starred.push(action.payload);
    },

    unStarRepo: (state, action: PayloadAction<number>) => {
      state.starred = state.starred.filter((id) => id !== action.payload);
    },
  },
});

export const { starRepo, unStarRepo } = githubSlice.actions;

export const selectStarredReposIds = (state: RootState) => state.github.starred;

export default githubSlice.reducer;
