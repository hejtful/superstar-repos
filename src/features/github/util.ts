export const buildGitHubApiQueryParam = <T>(params: T) => {
  const paramKeys = Object.keys(params) as Array<keyof T>;

  return paramKeys.reduce((queryString, currentParamKey) => {
    const currentParamValue = params[currentParamKey];

    if (!currentParamValue) return queryString;

    return queryString.concat(`${currentParamKey}:"${currentParamValue}"`);
  }, '');
};
