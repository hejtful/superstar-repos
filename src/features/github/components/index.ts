import { LanguageFilter } from './LanguageFilter';
import { RepoListItem } from './RepoListItem';

export { LanguageFilter, RepoListItem };
