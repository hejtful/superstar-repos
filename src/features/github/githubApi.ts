import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

import { GetReposResponse, Repository, RepoApiQuery } from './githubTypes';
import { buildGitHubApiQueryParam } from './util';
import { languagesPerPage } from './constants';

export const githubApi = createApi({
  reducerPath: 'githubApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.github.com/' }),
  endpoints: (builder) => ({
    getRepos: builder.query<Repository[], RepoApiQuery>({
      query: (params) => ({
        url: `search/repositories`,
        params: {
          q: buildGitHubApiQueryParam<RepoApiQuery>(params),
          sort: `stars`,
          order: `desc`,
          page: 1,
          per_page: languagesPerPage,
        },
      }),
      transformResponse: (response: GetReposResponse) => response.items,
    }),
  }),
});

export const { useGetReposQuery } = githubApi;
