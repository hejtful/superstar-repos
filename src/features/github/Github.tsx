import { useState } from 'react';

import { getLastWeekQueryString } from '../../app/util';
import { useGetReposQuery } from './githubApi';
import { LanguageFilter, RepoListItem } from './components/';

import styles from './Github.module.css';

const now = new Date();
const created = getLastWeekQueryString(now);

export const errorMessage = 'Something went wrong. Please try again later';
export const loadingMessage = 'Loading...';

export function Github() {
  const [language, setLanguage] = useState<string | null>(null);
  const {
    data: repos,
    error,
    isLoading,
  } = useGetReposQuery({ created, language });

  return (
    <main className={styles.main}>
      <div className={styles.banner}>
        <h1 className={styles.bannerTitle}>Superstar Repos</h1>
        <p className={styles.bannerDescription}>
          See the repositories that the GitHub community starred the most in the
          previous week
        </p>
      </div>

      <div className={styles.container}>
        <div className={styles.wrapper}>
          <div className={styles.actions}>
            <LanguageFilter
              onChange={(event) =>
                setLanguage((event.target as HTMLSelectElement).value)
              }
            />
          </div>

          <div>
            {error ? (
              // Error message
              <div className={styles.statusMessage}>{errorMessage}</div>
            ) : isLoading ? (
              // Loading message
              <div className={styles.statusMessage}>{loadingMessage}</div>
            ) : repos?.length ? (
              // Loaded repos
              repos.map((repo) => (
                <RepoListItem
                  key={repo.id}
                  repo={repo}
                  isStarred={false}
                  onStarButtonClick={() => ({})}
                  onUnStarButtonClick={() => ({})}
                />
              ))
            ) : null}
          </div>
        </div>
      </div>
    </main>
  );
}
