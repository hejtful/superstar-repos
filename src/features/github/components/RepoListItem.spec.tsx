import { render } from '@testing-library/react';
import { mockRepositories } from '../../../test/mocks/github/repositories';

import { RepoListItem } from './RepoListItem';

it('should render properly when not starred', () => {
  const { asFragment } = render(
    <RepoListItem
      repo={mockRepositories[0]}
      isStarred={false}
      onStarButtonClick={jest.fn()}
      onUnStarButtonClick={jest.fn()}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('should render properly when starred', () => {
  const { asFragment } = render(
    <RepoListItem
      repo={mockRepositories[1]}
      isStarred={true}
      onStarButtonClick={jest.fn()}
      onUnStarButtonClick={jest.fn()}
    />
  );

  expect(asFragment()).toMatchSnapshot();
});
