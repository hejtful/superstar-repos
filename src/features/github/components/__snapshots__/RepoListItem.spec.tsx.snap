// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly when not starred 1`] = `
<DocumentFragment>
  <article
    class="wrapper"
    data-testid="repo-list-item"
  >
    <div
      class="titleWrapper"
    >
      <h1
        class="title"
      >
        <a
          class="link"
          href="https://github.com/smartvotedocs/votesmart2021"
        >
          smartvotedocs
          <span
            class="repoName"
          >
             / votesmart2021
          </span>
        </a>
      </h1>
      <button
        class="button"
        type="button"
      >
        Star
      </button>
    </div>
    <p
      class="description"
    >
      Кандидаты Умного голосования в Госдуму 2021
    </p>
    <div
      class="details"
    >
      <span>
        Language: Unknown
      </span>
      <span>
        Stars: 1342
      </span>
    </div>
  </article>
</DocumentFragment>
`;

exports[`should render properly when starred 1`] = `
<DocumentFragment>
  <article
    class="wrapper"
    data-testid="repo-list-item"
  >
    <div
      class="titleWrapper"
    >
      <h1
        class="title"
      >
        <a
          class="link"
          href="https://github.com/phoboslab/q1k3"
        >
          phoboslab
          <span
            class="repoName"
          >
             / q1k3
          </span>
        </a>
      </h1>
      <button
        class="button"
        type="button"
      >
        Un-star
      </button>
    </div>
    <p
      class="description"
    >
      A tiny FPS for js13k
    </p>
    <div
      class="details"
    >
      <span>
        Language: JavaScript
      </span>
      <span>
        Stars: 1053
      </span>
    </div>
  </article>
</DocumentFragment>
`;
